### ISSUE TITLE
API reformulation

### ISSUE THREAD

**Issue by MartinPerez** [nilearn/nilearn#2369]
A recent discussion pointed at several current weaknesses of the API:
- The error prone need to reinput parameters. For example to instantiate design matrices at different pipeline steps.
- Lack of separation between data and parameters. Like the relationship between TR and design. We might want to separate design_spec from design_matrix.
- Misnaming of certain functions in the analysis process. Like the transform in the GLM object that can be misleading. Other functions might be added to the GLM object to get certain types of ouputs.
- Too much historical baggage. We might want to experiment with different conventions. Like replacing model specification with contrast vectors by an R/Statsmodels type of specification.

The importance of getting the second level analysis object right was emphasized. There were some issues raised related to mask application at first or second level and how to handle complex designs (ex. factorial design). We have to balance carefully user surprisal with limiting the power of the framework for mere historical reasons.

Here a prototype proposal of API changes:

paradigm = load_paradigm('paradigm_sub1.csv')
confounds = pandas.read_csv('motion.csv')

model = GLM(hrf_model="spm", drift_model="", t_r=None,
            mask_img=False)
model.fit(paradigm, epi_imgs, confounds)
print(model.summary())
model.compute_contrast('seen - unseen')

second_level = SecondLevel(type="ffx", mask='auto')
second_level.fit([model1, model2, model3], design=None)
second_level.compute_contrast('(seen - unseen) \* (schizo - control)')
second_level.compute_contrast(first='(seen - unseen)'
                              second='(schizo - control)')

**Comment by bthirion** [comment]
Closing for now. Some of it is deprecated, although we're not yet fully happy with nilearn.glm API.
